core/src/main/java/org/infinispan/cache/impl/EncoderCache.java

```diff
@@ -0,0 +1,878 @@
+   }
+
+   @Override
+   public CompletableFuture<Void> putAllAsync(Map<? extends K, ? extends V> data, long lifespan, TimeUnit lifespanUnit, long maxIdle, TimeUnit maxIdleUnit) {
+      return super.putAllAsync(transcodeMapForWrite(data), lifespan, lifespanUnit, maxIdle, maxIdleUnit);
+   }
+
+   @Override
+   public AdvancedCache<K, V> withFlags(Flag... flags) {
+      AdvancedCache<K, V> returned = super.withFlags(flags);
+      if (returned != this && returned instanceof EncoderCache) {
+         EncoderCache returned1 = (EncoderCache) returned;
+         returned1.entryFactory = this.entryFactory;
+         returned1.encodingType = this.encodingType;
+         returned1.wrapper = this.wrapper;
+         returned1.encoder = encoder;
+      }
+      return returned;
+   }
+
+   @Override
+   public AdvancedCache<K, V> withSubject(Subject subject) {
+      AdvancedCache<K, V> returned = super.withSubject(subject);
+      if (returned != this && returned instanceof EncoderCache) {
+         EncoderCache returned1 = (EncoderCache) returned;
+         returned1.entryFactory = this.entryFactory;
+         returned1.encodingType = this.encodingType;
+         returned1.wrapper = this.wrapper;
+         returned1.encoder = this.encoder;
+      }
+      return returned;
+   }
+
+   @Override
+   public CompletableFuture<V> putIfAbsentAsync(K key, V value) {
+      return super.putIfAbsentAsync(keyToStorage(key), valueToStorage(value)).thenApply(transcodeValueForRead);
+   }
+
+   @Override
+   public CompletableFuture<V> putIfAbsentAsync(K key, V value, long lifespan, TimeUnit unit) {
+      return super.putIfAbsentAsync(keyToStorage(key), valueToStorage(value), lifespan, unit).thenApply(transcodeValueForRead);
+   }
+
+   @Override
+   public boolean lock(K... keys) {
+      K[] transcoded = (K[]) Arrays.stream(keys).map(this::keyToStorage).toArray();
+      return super.lock(transcoded);
+   }
+
+   @Override
+   public boolean lock(Collection<? extends K> keys) {
+      return super.lock(transcodeKeysForWrite(keys));
+   }
+
+   @Override
+   public CompletableFuture<V> putIfAbsentAsync(K key, V value, long lifespan, TimeUnit lifespanUnit, long maxIdle, TimeUnit maxIdleUnit) {
+      return super.putIfAbsentAsync(keyToStorage(key), valueToStorage(value), lifespan, lifespanUnit, maxIdle, maxIdleUnit).thenApply(transcodeValueForRead);
+   }
+
+   @Override
+   public CompletableFuture<V> removeAsync(Object key) {
+      return super.removeAsync(keyToStorage(key)).thenApply(transcodeValueForRead);
+   }
+
+   @Override
+   public CompletableFuture<Boolean> removeAsync(Object key, Object value) {
+      return super.removeAsync(keyToStorage(key), valueToStorage(value));
+   }
+
+   @Override
+   public AdvancedCache<K, V> with(ClassLoader classLoader) {
+      AdvancedCache<K, V> returned = super.with(classLoader);
+      if (returned != this && returned instanceof EncoderCache) {
+         ((EncoderCache) returned).entryFactory = this.entryFactory;
```

Comment:
You are missing some variables here. I wonder if we should make a common method for configuring these so if we add any more 'with' methods that they can just invoke the proper method instead of adding each of these lines.